threads.go

```diff
@@ -555,42 +578,51 @@ func (t *threads) putRecord(ctx context.Context, id thread.ID, lid peer.ID, rec
 		return err
 	}
 
-	// Save the record locally
-	// Note: These get methods will return cached nodes.
-	block, err := rec.GetBlock(ctx, t)
-	if err != nil {
-		return err
-	}
-	event, ok := block.(*cbor.Event)
-	if !ok {
-		return fmt.Errorf("invalid event")
-	}
-	header, err := event.GetHeader(ctx, t, nil)
-	if err != nil {
-		return err
-	}
-	body, err := event.GetBody(ctx, t, nil)
-	if err != nil {
-		return err
-	}
-	err = t.AddMany(ctx, []format.Node{rec, event, header, body})
-	if err != nil {
-		return err
-	}
+	for i := len(unknownRecords) - 1; i >= 0; i-- {
+		r := unknownRecords[i]
+		// Save the record locally
+		// Note: These get methods will return cached nodes.
+		block, err := r.GetBlock(ctx, t)
+		if err != nil {
+			return err
+		}
+		event, ok := block.(*cbor.Event)
+		if !ok {
+			event, err = cbor.EventFromNode(block)
+			if err != nil {
+				return fmt.Errorf("invalid event: %v", err)
+			}
+		}
+		header, err := event.GetHeader(ctx, t, nil)
+		if err != nil {
+			return err
+		}
+		body, err := event.GetBody(ctx, t, nil)
+		if err != nil {
+			return err
+		}
+		err = t.AddMany(ctx, []format.Node{r, event, header, body})
+		if err != nil {
+			return err
+		}
+
+		log.Debugf("put record %s (thread=%s, log=%s)", r.Cid().String(), id, lg.ID)
 
+		// Notify local listeners
+		err = t.bus.SendWithTimeout(&record{
+			Record:   r,
+			threadID: id,
+			logID:    lg.ID,
+		}, time.Second)
+		if err != nil {
+			return err
+		}
+	}
 	// Update head
-	if err = t.store.SetHead(id, lg.ID, rec.Cid()); err != nil {
+	if err = t.store.SetHead(id, lg.ID, unknownRecords[0].Cid()); err != nil {
```

Comment:
After looping for all the unknown records in order (reverse slice order), we set the current head to the first element (which is always the one we received as a param in the method).

If you recall a previous comment, this is the ### mark.